### PR TITLE
@storybook/addon-knobs: add framework-specific entry points

### PR DESCRIPTION
**Pull request by joscha**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/storybooks/storybook/pull/1832 and https://github.com/storybooks/storybook/blob/master/MIGRATION.md#refactored-knobs
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

The Storybook knobs add-on has changed its API due to the support of more frameworks. The old way of importing from `@storybook/addon-knobs` is still available but deprecated. The right way (without performance penalty of importing all frameworks) is to import the specific version, e.g.:

* In the case of React: `import { withKnobs, text, boolean, number } from '@storybook/addon-knobs/react';`
* In the case of Vue: `import { ... } from '@storybook/addon-knobs/vue';`
* In the case of Angular: `import { ... } from '@storybook/addon-knobs/angular';`

Locally, for TS this is currently broken due to the changes in knobs, I added a d.ts file locally:
```
declare module '@storybook/addon-knobs/react' {
  export * from '@storybook/addon-knobs';
}
```
which works there, but I am unsure about this approach within DefinitelyTyped, @martynaskadisa @andy-ms any ideas?

### CODE REVIEW

**Review by ndelangen — APPROVED** [review]

**Review by martynaskadisa — APPROVED** [review]
@joscha I like this approach, maybe someone from maintainers could confirm if this is the right way.

File: types/storybook__addon-knobs/angular.d.ts
Comment (andy-ms) [inline types/storybook__addon-knobs/angular.d.ts]: Avoid `declare module` and just write exports directly at the top-level. (The entire file could be just `export * from "./index";`.

File: types/storybook__addon-knobs/index.d.ts
Comment (andy-ms) [inline types/storybook__addon-knobs/index.d.ts]: Is this react-specific? Should we avoid re-exporting this from `angular.d.ts` and `vue.d.ts`?

File: types/storybook__addon-knobs/index.d.ts
Comment (andy-ms) [inline types/storybook__addon-knobs/index.d.ts]: These shouldn't be necessary if these files are re-written to avoid `declare module` (see above comment)

**Review by ghost — CHANGES_REQUESTED** [review]

File: types/storybook__addon-knobs/react.d.ts
Comment (andy-ms) [inline types/storybook__addon-knobs/react.d.ts]: Remember to add these to "files" in `tsconfig.json`.

File: types/storybook__addon-knobs/angular.d.ts
Comment (joscha) [reply types/storybook__addon-knobs/angular.d.ts]: done in a65fd408d25c20942f489d136079bef4ba879518

File: types/storybook__addon-knobs/index.d.ts
Comment (joscha) [reply types/storybook__addon-knobs/index.d.ts]: I believe because this is the plugin function, the signature is still correct - @ndelangen can you clarify, please?

File: types/storybook__addon-knobs/index.d.ts
Comment (joscha) [reply types/storybook__addon-knobs/index.d.ts]: a65fd408d25c20942f489d136079bef4ba879518

File: types/storybook__addon-knobs/react.d.ts
Comment (joscha) [reply types/storybook__addon-knobs/react.d.ts]: a65fd408d25c20942f489d136079bef4ba879518

File: types/storybook__addon-knobs/tsconfig.json
Comment (andy-ms) [inline types/storybook__addon-knobs/tsconfig.json]: `bue` -> `vue`

Comment (joscha) [reply types/storybook__addon-knobs/tsconfig.json]: 🤦‍♂️ fixed in ce620ace2855b7e4a82bf818aa9890a547d46bcc

**Review by ghost — CHANGES_REQUESTED** [review]
Let's wait to hear about `withKnobs` before merging

File: types/storybook__addon-knobs/index.d.ts
Comment (joscha) [reply types/storybook__addon-knobs/index.d.ts]: @ndelangen @hypnosphi @shilman could you shed some light on this, please?

**Comment by joscha** [comment]
@andy-ms given that noone else from the Storybook team answers and the add-on is unusable currently, do you think we can get this in for now and fix up later in case the signature with the react-specific render function is incorrect? At least that would unblock usage.

File: types/storybook__addon-knobs/index.d.ts
Comment (Hypnosphi) [reply types/storybook__addon-knobs/index.d.ts]: One should look at implementations to figure out the return types. See vue for instance https://github.com/storybooks/storybook/blob/master/addons/knobs/src/vue/index.js

Comment (andy-ms) [reply types/storybook__addon-knobs/index.d.ts]: >  so the signature should be framework-dependent

I think this could be done by making a different version of the function in each of `angular.d.ts`, `react.d.ts`, and `vue.d.ts`.

**Comment by ndelangen** [comment]
I'm unsure what the question is, that this is blocked on?

Perhaps @igor-dv can help?

**Comment by joscha** [comment]
@ndelangen, no worries @hypnosphi answered a couple of hours ago, will
adapt the PR!

On 2 Feb 2018 19:03, "Norbert de Langen" <notifications@github.com> wrote:

I'm unsure what the question is, that this is blocked on?

Perhaps @igor-dv <https://github.com/igor-dv> can help?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/DefinitelyTyped/DefinitelyTyped/pull/22939#issuecomment-362513884>,
or mute the thread

**Comment by mhegazy** [comment]
@andy-ms is this PR good to go?

**Pull request merged**
